### ISSUE TITLE
Collection.Append() with open generic and lifestyle throws an exception

### ISSUE THREAD

**Issue by henriblMSFT** [simpleinjector/SimpleInjector#857]
Attempting to append an open generic to a collection while specifying a lifestyle throws exception. 

Verified with SimpleInjector 4.10.2 and 5.0.3

I would expect this scenario to be supported.

The following is currently supported
```csharp
container.Collection.Append(
    typeof(IEventHandler<>),
    typeof(NewConstraintEventHandler<>));
```
While using the overload with a Lifestyle supplied throws an exception
```csharp
container.Collection.Append(
    typeof(IEventHandler<>),
    typeof(NewConstraintEventHandler<>),
    Lifestyle.Singleton);
```
An `ArgumentException` is thrown with the following message: 
```
The supplied type NewConstraintEventHandler<TEvent> is an open-generic type. 
This type cannot be used for registration using this method.
```

callstack:
```
Requires.IsNotOpenGenericType(Type type, String paramName) line 54
Lifestyle.CreateRegistration(Type concreteType, Container container) line 606
ContainerCollectionRegistrator.Append(Type serviceType, Type implementationType, Lifestyle lifestyle) line 491
```

**Comment by dotnetjunkie** [comment]
I can confirm this is a bug. Let me do some analysis and get back to you.

**Comment by dotnetjunkie** [comment]
Hi @henriblMSFT,

After investigating the code base, I came to the conclusion that `Append` has never supported the registration of open generic types with a lifestyle. This, however, is a feature that IMO _should_ be supported. Unfortunately, this isn't a simple fix, so a better analysis must be done in order to introduce accidental breaking changes.

I'll keep this issue open and will try to add this feature in a coming release. 

In the meantime, you can use the following workaround:

``` c#
container.Collection.Append(typeof(IEventHandler<>), typeof(NewConstraintEventHandler<>));

container.Register(
    typeof(NewConstraintEventHandler<>),
    typeof(NewConstraintEventHandler<>),
    Lifestyle.Singleton);
```

This works, because the collection subsystem will peak at existing registrations before trying to create a registration of its own. It will, therefore, use the registration made using `Register`. This is by design.

I will update this issue when there is more information about progress.

**Comment by dotnetjunkie** [comment]
Branch [bug-857](https://github.com/simpleinjector/SimpleInjector/tree/bug-857) created.

**Comment by dotnetjunkie** [comment]
Update: The same problem holds for the `Collection.Register` method. The following yields the same exception:

``` c#
container.Collection.Register(
    typeof(IEventHandler<>),
    new[] { typeof(NewConstraintEventHandler<>), typeof(StructConstraintEventHandler<>) },
    Lifestyle.Singleton);
```

**Comment by dotnetjunkie** [comment]
This feature is now part of the just-released v5.1.
